### PR TITLE
raft: add read index for RawNode

### PR DESCRIPTION
**Pull request by hhkbp2**
Hi,

This PR adds read index(PR #5553) for RawNode, and test case.

### CODE REVIEW

**Comment by xiang90** [comment]
I need to hold on this pr. We need to merge #6275 first.

**Comment by xiang90** [comment]
#6275 is merged. The interface is changed. Can you change this pr accordingly?

**Comment by hhkbp2** [comment]
@xiang90 Sure.

**Comment by hhkbp2** [comment]
PTAL @xiang90

File: raft/rawnode_test.go [hunk 1]
```diff
@@ -110,6 +110,68 @@ func TestRawNodeProposeAndConfChange(t *testing.T) {
...
+// to the underlying raft. It also ensures that ReadState can be read out.
+func TestRawNodeReadIndex(t *testing.T) {
+	msgs := []raftpb.Message{}
+	appendStep := func(r *raft, m raftpb.Message) {
+		msgs = append(msgs, m)
+	}
+	wrs := []ReadState{{Index: uint64(1), RequestCtx: []byte("somedata")}}
+
+	s := NewMemoryStorage()
+	var err error
```
Comment (xiang90) [inline raft/rawnode_test.go]: remove this variable? seems to be unnecessary.

File: raft/rawnode_test.go [hunk 2]
```diff
@@ -110,6 +110,68 @@ func TestRawNodeProposeAndConfChange(t *testing.T) {
...
+	c := newTestConfig(1, nil, 10, 1, s)
+	c.CheckQuorum = true
```
Comment (xiang90) [inline raft/rawnode_test.go]: remove check q? readindex will not rely on checkq by default.

File: raft/rawnode_test.go [hunk 3]
```diff
@@ -110,6 +110,68 @@ func TestRawNodeProposeAndConfChange(t *testing.T) {
...
+	rawNode, err := NewRawNode(c, []Peer{{ID: 1}})
+	if err != nil {
+		t.Fatal(err)
+	}
+	rawNode.raft.readStates = wrs
+	// ensure the ReadStates can be read out
```
Comment (xiang90) [inline raft/rawnode_test.go]: this code block seems to be unnecessary. for this test, we should not care about if we can read out ready. we should only care about if the readindex workflow is correct.

Comment (hhkbp2) [inline raft/rawnode_test.go]: This test case maps the test case in `node_test.TestNodeReadIndex()`, which checks the ready can be read out.

File: raft/rawnode_test.go (hunk 2 above)
Comment (hhkbp2) [inline raft/rawnode_test.go]: Address comment.

File: raft/rawnode_test.go (hunk 3 above)
Comment (xiang90) [inline raft/rawnode_test.go]: ok. fine with me then.

**Comment by xiang90** [comment]
Let's squash the commits into one. Then it would be ready to merge.

**Comment by xiang90** [comment]
@hhkbp2 Thanks a lot!

**Pull request merged**